[webkitpy] Add --udids option to run-webkit-tests.

bls1999 opened the pull request:
#### c6e2256db28c86d38e380e29bb8d12a69c79dfdf
<pre>
[webkitpy] Add --udids option to run-webkit-tests.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282198">https://bugs.webkit.org/show_bug.cgi?id=282198</a>
<a href="https://rdar.apple.com/138779069">rdar://138779069</a>

Reviewed by Jonathan Bedard.

This change adds a method of specifying the desired simulator to use to run
layout tests.

The new flow of simulated_device.py:
1. Try to find/boot simulators matching the specified UDID(s) (specify more
    than one by separating with a comma).
2. If device requests remain after handling the devices specified by UDID, try
    to attach to other booted simulators.
3. If we still have device requests after utilizing all available and matching
    existing simulators, create and boot simulators to match the requests.

* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py: Add --udids option to `parse_args`.
* Tools/Scripts/webkitpy/port/device_port.py: Pass specified UDIDs to the device manager.
* Tools/Scripts/webkitpy/xcode/simulated_device.py: SimulatedDeviceManager changes to accept and handle new data.
    -&gt; (initialize_devices): Change flow to above.
    -&gt; (_does_fulfill_request): Add option to include non-booted simulators.
    -&gt; (_validate_running_device_against_requests): Move requests validation from `initialize_devices`.

Canonical link: <a href="https://commits.webkit.org/285830@main">https://commits.webkit.org/285830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28ac23b18c1eae60a65c12473b994f723d0b4860

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78225 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25129 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76008 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58098 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16464 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76958 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63567 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38497 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/73399 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45060 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21056 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23462 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66612 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79780 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/626 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66425 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1351 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63580 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65704 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-default-size (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16277 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9595 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7773 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1172 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1201 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -420,30 +436,44 @@ def initialize_devices(cls, requests, host=None, name_base='Managed', simulator_
             requests = [requests]
 
+        # Parse user-specified UDIDs
+        if type(udids) is str:
```
To make things easier (avoid using commas and conditional splitting), you could specify `nargs='+'` in `optparse.make_option`.

bls1999 replied:
This is a good point! I was worried about messing up the positional argument flow. Turns out, this [isn't even be possible](https://stackoverflow.com/a/1025232) with `optparse`... :(

briannafan commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -420,30 +436,44 @@ def initialize_devices(cls, requests, host=None, name_base='Managed', simulator_
-            # DeviceRequests are compared by reference
-            requests_copy = [request for request in requests]
+            device_is_booted = device.platform_device.is_booted_or_booting()
+            if udids is not None and device.platform_device.udid not in udids:
```
nit: `if udids` should suffice - same for line 465

bls1999 replied:
Thanks for the tip! Resolved in latest commit.

briannafan commented on Tools/Scripts/webkitpy/xcode/simulated_device.py:
```diff
@@ -420,30 +436,44 @@ def initialize_devices(cls, requests, host=None, name_base='Managed', simulator_
+        # If the user passed UDIDs, we should check for matches among remaining booted devices and
+        # then resort to using any available ones or creating new ones that match the requests.
+        if udids is not None and len(deferred_booted_devices) and len(requests):
+            for matched_request, device in deferred_booted_devices:
```
If I understand correctly, if a device doesn't match, `matched_request = cls._does_fulfill_request(device, requests, True) = None`? And you indiscriminately add booted devices to the deferred list in line 454.

If you add a non-matching request to your deferred list and iterate over them here, do you want to be treating non-matches the same as matches? (Just checking if this behavior is intended, since I don't understand the full process.)

bls1999 replied:
Discussed offline, @briannafan didn't see this above:

```
if matched_request is None:
    continue
```

bls1999 committed:
[webkitpy] Add --udids option to run-webkit-tests.
https://bugs.webkit.org/show_bug.cgi?id=282198
rdar://138779069

Reviewed by Jonathan Bedard.

This change adds a method of specifying the desired simulator to use to run
layout tests.

The new flow of simulated_device.py:
1. Try to find/boot simulators matching the specified UDID(s) (specify more
    than one by separating with a comma).
2. If device requests remain after handling the devices specified by UDID, try
    to attach to other booted simulators.
3. If we still have device requests after utilizing all available and matching
    existing simulators, create and boot simulators to match the requests.

* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py: Add --udids option to `parse_args`.
* Tools/Scripts/webkitpy/port/device_port.py: Pass specified UDIDs to the device manager.
* Tools/Scripts/webkitpy/xcode/simulated_device.py: SimulatedDeviceManager changes to accept and handle new data.
    -> (initialize_devices): Change flow to above.
    -> (_does_fulfill_request): Add option to include non-booted simulators.
    -> (_validate_running_device_against_requests): Move requests validation from `initialize_devices`.

Canonical link: https://commits.webkit.org/285830@main
